File: modify_product.dart
import 'dart:convert';
import 'package:barcode_scan2/platform_wrapper.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

import '../utils/primary_button.dart';
import '../utils/primary_textfield.dart';

class ModifyProduct extends StatefulWidget {
  const ModifyProduct({Key? key}) : super(key: key);

  @override
  State<ModifyProduct> createState() => _ModifyProductState();
}

class _ModifyProductState extends State<ModifyProduct> {
  TextEditingController barcodeController = TextEditingController();
  TextEditingController productController = TextEditingController();
  TextEditingController priceController = TextEditingController();
  TextEditingController brandController = TextEditingController();
  TextEditingController departmentController = TextEditingController();
  TextEditingController measureController = TextEditingController();

  bool isProductFetched = false; // To track if product details are fetched

  String scannedBarcode = 'No barcode scanned';

  // Function to start barcode scanning
  Future<void> scanBarcode() async {
    try {
      var result = await BarcodeScanner.scan();  // Start the barcode scanning
      if (result.rawContent.isNotEmpty) {
        setState(() {
          barcodeController.text = result.rawContent;  // Update the barcode text
          _fetchProductDetails();
        });
      }
    } catch (e) {
      setState(() {
        scannedBarcode = 'Failed to scan barcode: $e';
      });
    }
  }

  // Fetch product details by barcode
  _fetchProductDetails() async {
    var barcode = barcodeController.text.trim();

    // Handle empty barcode
    if (barcode.isEmpty) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('Please enter or scan a barcode')),
      );
      return;
    }

    var url = Uri.parse(
        'http://endeers.com.preview.services/get_product.php?barcode=$barcode');

    try {
      var response = await http.get(url);

      if (response.statusCode == 200) {
        var productData = json.decode(response.body);

        if (productData.containsKey('error')) {
          // Show error message if product not found
          ScaffoldMessenger.of(context)
              .showSnackBar(SnackBar(content: Text(productData['error'])));
        } else {
          // Populate the text fields with the product data
          setState(() {
            productController.text = productData['product_name'] ?? 'Unknown';
            priceController.text = productData['price']?.toString() ?? '0';
            brandController.text = productData['brand'] ?? 'Unknown';
            measureController.text = productData['measure'] ?? 'Unknown';
            departmentController.text = productData['department'] ?? 'Unknown';
          });
        }
      } else {
        // Handle error response
        ScaffoldMessenger.of(context).showSnackBar(
            const SnackBar(content: Text('Failed to fetch product details')));
        print('Error: ${response.body}');
      }
    } catch (error) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Error: $error')),
      );
      print('Exception: $error');
    }
  }

  Future<void> _modifyProduct() async {
    var url = Uri.parse('http://endeers.com.preview.services/modify_product.php'); // Your modify product endpoint
    try {
      var response = await http.post(url, body: {
        "barcode": barcodeController.text,
        "product_name": productController.text,
        "price": priceController.text,
        "brand": brandController.text,
        "department": departmentController.text,
        "measure": measureController.text
      });

      var jsonResponse = jsonDecode(response.body);

      if (response.statusCode == 200 && jsonResponse['success'] == true) {
        ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(content: Text(jsonResponse['message'] ?? 'Product modified successfully!'))
        );
      } else {
        ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(content: Text('Failed to modify product: ${jsonResponse['message']}'))
        );
      }
    } catch (e) {
      ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text('Error: $e'))
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: IconButton(
          icon: Icon(Icons.arrow_back_ios_new_rounded),
          color: Colors.blueGrey,
          onPressed: () {
            Navigator.pop(context);
          },
        ),
        centerTitle: true,
        title: const Text(
          'Modify Product',
          style: TextStyle(
              fontSize: 25, color: Colors.blueGrey, fontWeight: FontWeight.bold),
        ),
        backgroundColor: Colors.transparent,
        elevation: 0,
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            Row(
              children: [
                Expanded(
                  child: PrimaryTextField('Bar Code', barcodeController),
                ),
                IconButton(
                  icon: const Icon(Icons.barcode_reader),
                  onPressed: () async {
                    scanBarcode();
                  },
                ),
                IconButton(
                  icon: const Icon(Icons.search),
                  onPressed: () async {
                    _fetchProductDetails();
                  },
                ),
              ],
            ),
            PrimaryTextField('Product', productController),
            PrimaryTextField('Price', priceController),
            PrimaryTextField('Brand', brandController),
            PrimaryTextField('Department', departmentController),
            PrimaryTextField('Measure', measureController),
            PrimaryButton('Modify Product', () async {
              await _modifyProduct();
            }),
          ],
        ),
      ),
    );
  }
}

File: modify_product.php
<?php
// Database credentials
$servername = "localhost";
$username = "endeersc_srn";
$password = "TrDduMGvva2ZDD9vdMH5";
$dbname = "endeersc_srn";

// Create connection
$conn = new mysqli($servername, $username, $password, $dbname);

// Check connection
if ($conn->connect_error) {
    die("Connection failed: " . $conn->connect_error);
}

// Check if data was posted
if ($_SERVER['REQUEST_METHOD'] == 'POST') {
    // Get the posted data
    $barcode = $_POST['barcode'];
    $product_name = $_POST['product_name'];
    $price = $_POST['price'];
    $brand = $_POST['brand'];
    $department = $_POST['department'];
    $measure = $_POST['measure'];

    // Ensure that barcode is provided
    if (empty($barcode)) {
        echo json_encode([
            "success" => false,
            "message" => "Barcode is required"
        ]);
        exit();
    }

    // Check if the product exists with the given barcode
    $check_sql = "SELECT * FROM products WHERE barcode = '$barcode'";
    $result = $conn->query($check_sql);

    if ($result->num_rows > 0) {
        // Update the product details
        $update_sql = "UPDATE products 
                       SET product_name = '$product_name', price = '$price', brand = '$brand', 
                           department = '$department', measure = '$measure' 
                       WHERE barcode = '$barcode'";

        if ($conn->query($update_sql) === TRUE) {
            echo json_encode([
                "success" => true,
                "message" => "Product modified successfully"
            ]);
        } else {
            echo json_encode([
                "success" => false,
                "message" => "Error updating product: " . $conn->error
            ]);
        }
    } else {
        echo json_encode([
            "success" => false,
            "message" => "Product with the given barcode not found"
        ]);
    }
} else {
    echo json_encode([
        "success" => false,
        "message" => "Invalid request"
    ]);
}

// Close the connection
$conn->close();
?>
